### PR TITLE
git-reset

### LINKED ISSUE #13: Add git reset
- soft vs. hard reset
- how to undo changes
- clean

### PR DESCRIPTION
**Pull request by graywolf**
fixes #13

### CODE REVIEW

File: 11-Reset/readme.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+git-reset
+=========
+
+> Reset current HEAD to the specified state
+
+Git reset has three forms. In all of the &lt;tree-ish&gt; defaults to HEAD.
+
+`git reset [-q] [<tree-ish>] [--] <paths>...`
```
Comment (pzavoral) [inline 11-Reset/readme.md]: Header?

File: 11-Reset/readme.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
+git-reset
+=========
+
+> Reset current HEAD to the specified state
+
+Git reset has three forms. In all of the &lt;tree-ish&gt; defaults to HEAD.
```
Comment (pzavoral) [inline 11-Reset/readme.md]: put "in all of them" on the end of the sentence

**Comment by voloda** [comment]
I found most usefull git reset --soft and --hard. Could you please add these to this to see the difference?

Sometimes I am using git reset --soft for quick squashing whole history into single new commit

File: 11-Reset/readme.md (hunk 1 above)
Comment (graywolf) [inline 11-Reset/readme.md]: done

File: 11-Reset/readme.md (hunk 2 above)
Comment (graywolf) [inline 11-Reset/readme.md]: and done

**Comment by graywolf** [comment]
@voloda I'm not sure what you mean. At the end of the file all modes are described..

**Comment by voloda** [comment]
@graywolf I know that soft/hard are 'described' there. Though - using git terminology wasn't much helpful for me. Behavior can be intuitive though having there some examples with real impact is IMHO more useful.

File: 11-Reset/readme.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+Git reset has three forms. &lt;tree-ish&gt; defaults to HEAD in all of them.
+
+git reset [-q] [<tree-ish>] [--] <paths>...
+-------------------------------------------
+
+> Can be thought of as opposite for `git-add`.
+
+Resets index entries for all &lt;paths&gt; to their state at &lt;tree-ish&gt;. Important part here is that this form of `git-reset` doesn't affect the working tree or current branch in any form.
+
+We can try it on our current repository. Use this form of reset to return to what `09-Blame/readme.md` looked like before it was added:
```
Comment (pzavoral) [inline 11-Reset/readme.md]: Do not use this GIT repository, create a special training repository for git-reset

File: 11-Reset/readme.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+> Reset current HEAD to the specified state
+
+Git reset has three forms. &lt;tree-ish&gt; defaults to HEAD in all of them.
+
+git reset [-q] [<tree-ish>] [--] <paths>...
+-------------------------------------------
+
+> Can be thought of as opposite for `git-add`.
+
+Resets index entries for all &lt;paths&gt; to their state at &lt;tree-ish&gt;. Important part here is that this form of `git-reset` doesn't affect the working tree or current branch in any form.
```
Comment (pzavoral) [inline 11-Reset/readme.md:15]: This is true only for soft reset, hard reset does change the working tree

File: 11-Reset/readme.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+git reset [<mode>] [<commit>]
+-----------------------------
+
+This form resets the current branch head to &lt;commit&gt; and depending on the mode also updates the index or working tree.
+
+* --soft
+	* does not touch the index or the working tree
+* --mixed (default)
+	* resets the index but not the working tree
+* --hard
```
Comment (pzavoral) [inline 11-Reset/readme.md:77]: I think hard reset is also useful - there should be a sample of hard reset as well

File: 11-Reset/readme.md (hunk 4 above)
Comment (graywolf) [inline 11-Reset/readme.md:15]: Please see the exact for of this reset, it doesn't have soft/hard modes. Resets with modes are in section bellow.

**Comment by voloda** [comment]
Perhaps extend the comment here to:
- Does not touch the index or the working tree. Your files will remain in their last state on the disk and ready for `git-commit`. This allows you to recreate set of changes as a new commit.

**Comment by voloda** [comment]
Also - it would be great to talk a bit about the depth of history:

`git reset --soft HEAD~3`

**Comment by graywolf** [comment]
but the "HEAD~3" part doesn't directly relate to this, should it be possible mentioned somewhere else?

File: 11-Reset/readme.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+git reset [<mode>] [<commit>]
+-----------------------------
+
+This form resets the current branch head to &lt;commit&gt; and depending on the mode also updates the index or working tree.
+
+* --soft
+	* does not touch the index or the working tree, your files will remain as they were and in "Changes to be commited" (index).
+	* this is useful if you've forgotten to add something into last commit
+
+	git reset --soft HEAD~1
```
Comment (voloda) [inline 11-Reset/readme.md]: Please format as:

```
> git reset --soft HEAD~1
> git add forgotten_file
> git commit
```

File: 11-Reset/readme.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	* all your files are the same, they are not touched, but they are not in index anymore, they show up in local changes / untracked files
+		* you need to git-add them again before commiting
+* --hard
+	* resets the index and working tree
+	* all changes to tracked files are discarded
+	* useful to throw away changes you just don't want to have
+		* you screwed up merge and want to try again?
+		* your code is completely wrong and you want to start over?
+
+	git reset --hard HEAD~1
```
Comment (voloda) [inline 11-Reset/readme.md]: Please format as

```
> git reset --hard HEAD~1
```

**Comment by voloda** [comment]
@graywolf `HEAD~2` - I use it mostly here. What are other places? Perhaps `git log`? Feel free to add it. I would though like to have it here as well...

**Comment by graywolf** [comment]
git log is one place, git diff other where I use it.. git diff another. or git checkout sometimes. but sure, I'll add it here

**Comment by graywolf** [comment]
@voloda could you please merge this? I think I resolved all feedback, but don't know as what number I should merge it, there is no hole left for it.. thx

**Comment by voloda** [comment]
:+1: I will resolve the naming

**Pull request merged**